doc/mgmt/SONiC_PIM_SM_HLD.md

```diff
@@ -0,0 +1,1036 @@
+
+### 3.1.2 ipmcorchagent changes
+As part of the PIM ASM support following functionalities attained by ipmcorchagent,<br> 
+1. CPU bit update for (*,G) and (S,G) routes<br> 
+2. Updating the KAT timer value for (S,G) routes from FRR to ASIC<br> 
+
+#### 3.1.2.1 CPU bit update
+
+CPU bit must be updated for (*,G) or (S,G) entry to punt the packets to FRR stack using CPU port. <br>
+To achieve this functionality during creation of IPMC entry in BCM we need to set the action as SAI_PACKET_ACTION_COPY  .<br>
+If CPU port removed from the (*,G) or (S,G) entry then ipmcorchagent need to set action to SAI_PACKET_ACTION_COPY_CANCEL <br>
+
+Following are the sequence of events handled in ipmcorchagent for CPU bit update,<br>
+1. ipmcfpmsyncd update APPL_DB with (S,G) or (*,G) and OIF as "pimreg" in APP_IPMC_ROUTE_TABLE_NAME table
+2. ipmcorchagent gets notified with APP_IPMC_ROUTE_TABLE_NAME<br>
+3. If operation is SET with ifname as "pimreg", set the ACTION as SAI_PACKET_ACTION_COPY
+4. If operation is DELETE with ifname as "pimreg", set the ACTIOn as SAI_PACKET_ACTION_COPY_CANCEL<br>
+
+| Existing Entry  |   New Entry | Operation | Action |
+| ----------------|-------------|-----------|--------|
+| None            | (*,G) or (S,G) OIF as pimreg | SET | SAI_PACKET_ACTION_COPY (IPMC entry created) |
+| (*,G) or (S,G) OIF as pimreg | (*,G) or (S,G) OIF as pimreg | DEL | IPMC entry removed |
+| None            | (*,G) or (S,G) OIF as Vlan10 | SET | SAI_PACKET_ACTION_FORWARD (IPMC entry created) |
+| (*,G) or (S,G) OIF as Vlan10 | (*,G) or (S,G) OIF as Vlan10,pimreg | SET | SAI_PACKET_ACTION_COPY |
+| (*,G) or (S,G) OIF as Vlan10,pimreg | (*,G) or (S,G) OIF as Vlan10 | SET | SAI_PACKET_ACTION_COPY_CANCEL |
+| (*,G) or (S,G) OIF as Vlan10,pimreg | (*,G) or (S,G) OIF as Vlan10,pimreg | DEL | IPMC entry removed |
+| (*,G) or (S,G) OIF as pimreg | (*,G) or (S,G) OIF as pimreg | DEL | IPMC entry removed |
+
+#### 3.1.2.2 IIF expansion in OIF List
+
+IIF member ports to be included in IPMC group member on following cases,
+1. L3 multicast case, local receiver on the IIF vlan then add IIF vlan in OIF list
```

Comment:
Not clear why this is needed, SAI has SAI_IPMC_ENTRY_ATTR_RPF_GROUP_ID object, the IIF would be matched to this and the ports from IIF(in case of VLAN) would go to IPMC l2 bitmap right? which module this happens, hope its not form common code.  if its inside NPU specific code its fine, otherwise it would be fine. or is it current SAI module does not have way to specify this?

Reply:
@mhampasagar  following are the discussion points with @ashok-krishnegowda on this requirement, and concluded on Approach1.
**Approach1:** 
IIF-VLAN will be always added as part of OIF-list by Orchagent
1.	If snooping is disabled, Orchagent will not fill L2MC-group-id for that IIF-VLAN.
          a.	SAI fills all VLAN-members today for these OIFs, but it fills it in L3-port-bitmap which modifies VLAN-tag.
          b.	In IIF-VLAN being OIF case, SAI should add VLAN-members as L2-port-bitmap. SAI should identify IIF-VLAN being OIF-VLAN by comparing RPF-group.
2.	If snooping is enabled, Orchagent will fill L2MC-group-id for that IIF-VLAN as like any other OIF-VLAN with snooped ports
 **Approach2:**
To use RPF group member  which is set to IIF. Here if snooping is disabled, RPF member will be the same data as filled today,   SAI need to include IIF vlan member ports to multicast group in L2 port bitmap.  If in IIF vlan, snooping is enabled, SAI_IPMC_GROUP_MEMBER_ATTR_EXTENSIONS_L2MC_OUTPUT_ID should be included in RPF Group member, so that orchagent can fill it and communicate to SAI.